app-ng/src/job.lisp

```diff
@@ -0,0 +1,145 @@
+    (format stream "QVM Job (~A)" (job-status-name job))))
+
+;; The %x-unsafe variants below do not acquire a lock.
+
+(defun %job-running-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    (job-status-running t)
+    (_ nil)))
+
+(defun job-running-p (job)
+  "Is the JOB in the RUNNING state."
+  (with-job-lock (job)
+    (%job-running-p-unsafe job)))
+
+(defun %job-killed-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    (job-status-killed t)
+    (_ nil)))
+
+(defun job-killed-p (job)
+  "Is the JOB in the KILLED state?"
+  (with-job-lock (job)
+    (%job-killed-p-unsafe job)))
+
+(defun %job-finished-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    ((job-status-finished _) t)
+    (_ nil)))
+
+(defun job-finished-p (job)
+  "Is the JOB in the FINISHED state?"
+  (with-job-lock (job)
+    (%job-finished-p-unsafe job)))
+
+(defun make-job (fn)
+  "Make a JOB object which will, under JOB-START, run FN in a thread.
+
+The returned JOB can be repeatedly started if the JOB-STATUS is FINISHED (or KILLED). Only interact with the job after acquiring the lock (or alternatively by using WITH-JOB-LOCK)."
+  (%make-job (lambda (job)
+               (adt:match job-status (job-status job)
+                 (job-status-fresh
+                  (setf (job-thread job)
+                        (bt:make-thread
+                         (lambda ()
+                           (let ((result 
+                                   (handler-case
+                                       (job-status-finished (funcall fn))
+                                     (error (c)
+                                       (job-status-error c)))))
+                             (with-job-lock (job)
+                               (setf (job-status job) result))))))
+                  (setf (job-status job) job-status-running))
+                 (_ (error 'job-error :job job :message "Job cannot be started because it is not fresh."))))))
+
+(defmacro define-job-action (action (job) &body body)
+  "Evaluate BODY with a lock held for JOB."
+  (multiple-value-bind (body decls docstr)
+      (alexandria:parse-body body :documentation t)
+    (declare (ignore decls))
+    `(defun ,action (,job)
+       ,docstr
+       (with-job-lock (,job)
+         ,@body))))
+
+;;; CUSTODIAL
+
+(define-job-action job-start (job)
+  "Start the JOB and return T. If JOB is already running, do nothing and return NIL."
+  (unless (%job-running-p-unsafe job)
+    (funcall (job-work-function job) job)
+    t))
+
+(defun force-kill-job (job)
+  "Forcefully kill JOB if running. Does not block and cannot be restarted."
+  (when (%job-running-p-unsafe job)
+    (bt:destroy-thread (job-thread job))
+    (setf (job-status job)
+          job-status-killed)))
+
+(define-job-action kill-job (job)
+  "Forcefully kill JOB if running as soon as lock is acquired. Job cannot be restarted."
+  (force-kill-job job))
+
+;;; sync/async
+
+(defun job-result (job)
```

Comment:
I don't think this is unsafe. This can be called concurrently and should be safe.

Reply:
same with job-peek

Reply:
I might be missing something. I'm imagining another thread might call `job-kill` right after `%job-running-p-unsafe` returns, then we attempt to `bt:join-thread` or something.

Reply:
@stylewarning @appleby consensus?

Reply:
I can see both sides. On the one hand, you don't want acquire the lock, then hold onto it for who-knows-how-long while you wait for `bt:join-thread` to return. Meanwhile, no-one can do anything to the job, not even kill it, since you're holding the lock. On the other hand, it is racy. We check the status to ensure it's running before calling `join-thread`, but nothing prevents another thread from changing the status in-between. Realistically, we should probably sleep on a condition variable waiting for the state transition running -> anything-else. But I'm fine with not overcomplicating things for now and dealing with it when/if it becomes a problem.

Reply:
i think appleby brings up good points and should be addressed.

killing a job should be considered very rare, usually called explicitly by a human and not part of some automated process